src/finn/custom_op/fpgadataflow/matrixvectoractivation.py

```diff
@@ -905,402 +965,68 @@ def generate_params(self, model, path):
 
+    def execute_node(self, context, graph):
+        node = self.onnx_node
+        in_act = context[node.input[0]]
+        mvau_w_init = [x for x in graph.initializer if x.name == node.input[1]][0]
+        mvau_w = np_helper.to_array(mvau_w_init)
+        # Matrix multiplication
+        if self.get_nodeattr("binaryXnorMode"):
+            # Note: activation/weights are expected to be binary (by design coming from the transformation inferring this operation mode)
+            result = xp.xnorpopcountmatmul(in_act, mvau_w)
+        elif (self.get_nodeattr("inputDataType") == "BIPOLAR" and self.get_nodeattr("weightDataType") == "BIPOLAR"):
+            result = xp.xnorpopcountmatmul((in_act+1)/2, (mvau_w+1)/2)
         else:
-            raise Exception(
-                """Please set mem_mode to "const", "decoupled", or external,
-                currently no other parameter value is supported!"""
-            )
-
-        # the threshold tensor is acc_type [PE][TMEM][N_THRES]
-        # partition for parallel access along PE and N_THRES
-        # dimensions (dims 1 and 3)
-        if self.calc_tmem() != 0:
-            # TODO find a better way of checking for no pregenerated thresholds
-            self.code_gen_dict["$PRAGMAS$"].append(
-                ("#pragma HLS ARRAY_PARTITION variable=threshs.m_thresholds " "complete dim=1")
-            )
-            self.code_gen_dict["$PRAGMAS$"].append(
-                ("#pragma HLS ARRAY_PARTITION variable=threshs.m_thresholds " "complete dim=3")
-            )
-            # add resource pragma for thresholds if set
-            if ram_style_thresholds == "distributed":
-                self.code_gen_dict["$PRAGMAS$"].append(
-                    ("#pragma HLS RESOURCE variable=threshs.m_thresholds " "core=ROM_2P_LUTRAM")
-                )
-            elif ram_style_thresholds == "block":
-                self.code_gen_dict["$PRAGMAS$"].append(
-                    ("#pragma HLS RESOURCE variable=threshs.m_thresholds " "core=ROM_2P_BRAM")
-                )
-            elif ram_style_thresholds == "auto":
-                # no pragma needed
-                pass
-            else:
-                raise Exception("Unrecognized ram_style_thresholds value:" + ram_style_thresholds)
+            result = np.matmul(in_act, mvau_w)
+        # Thresholding if noActivation==0
+        if self.get_nodeattr("noActivation") == 0:
+            mvau_thr_init = [x for x in graph.initializer if x.name == node.input[2]][0]
+            mvau_thr = np_helper.to_array(mvau_thr_init)
+            odt_is_bipolar = self.get_nodeattr("outputDataType") == DataType["BIPOLAR"]
+            out_scale = 2 if odt_is_bipolar else 1
+            out_bias = -1 if odt_is_bipolar else self.get_nodeattr("ActVal")
+            result = multithreshold(result, mvau_thr, out_scale, out_bias)
+        
+        context[node.output[0]] = result
 
     def code_generation_ipi(self):
```

Comment:
This can be split into an HLS- and RTL-specific part and be moved into the HLS or RTL custom op files.